sdks/python/apache_beam/yaml/yaml_provider_unit_test.py

```diff
@@ -0,0 +1,743 @@
+    """Test that the spec must contain the jar path when parsing java jar."""
+    spec = {'type': 'javaJar', 'transforms': self._urns}
+    with self.assertRaises(ValueError) as e:
+      ExternalProvider.provider_from_spec(spec)
+    self.assertRegex(
+        str(e.exception), ".*Unable to instantiate provider.*missing.*jar.*")
+
+  @mock.patch('urllib.parse.urlparse', return_value=ParseResult)
+  def test_java_jar_provider_from_spec(self, mock_url_parser):
+    """Test that provider is successfully parsed when spec is valid."""
+    spec = {
+        'type': 'javaJar',
+        'config': {
+            'jar': 'path/to/external.jar'
+        },
+        'transforms': self._urns
+    }
+    provider = ExternalProvider.provider_from_spec(spec)
+    self.java_provider_from_spec_asserts(provider)
+    mock_url_parser.assert_called_once()
+
+  def test_beam_jar_provider_from_spec_service_fails_without_gradle_target(
+      self):
+    """Test that the spec must contain the gradle target when parsing beam
+    jar."""
+    spec = {
+        'type': 'beamJar',
+        'config': {
+            'version': '1.0.0'
+        },
+        'transforms': self._urns
+    }
+    with self.assertRaises(ValueError):
+      ExternalProvider.provider_from_spec(spec)
+
+  def test_beam_jar_provider_from_spec(self):
+    """Test that provider is successfully parsed when spec is valid."""
+    spec = {
+        'type': 'beamJar',
+        'config': {
+            'gradle_target': 'path:to:external:jar', 'version': '1.0.0'
+        },
+        'transforms': self._urns
+    }
+    provider = ExternalProvider.provider_from_spec(spec)
+    self.java_provider_from_spec_asserts(provider)
+
+  def test_maven_jar_provider_from_spec_service_fails_without_required_args(
+      self):
+    """Test that the spec must contain the required args when parsing maven
+    jar."""
+    spec = {'type': 'mavenJar', 'transforms': self._urns}
+    with self.assertRaises(ValueError) as e:
+      ExternalProvider.provider_from_spec(spec)
+    self.assertRegex(
+        str(e.exception),
+        ".*Unable to instantiate provider.*"
+        "missing.*keyword-only arguments.*"
+        "artifact_id.*group_id.*version.*")
+
+  def test_maven_jar_provider_from_spec(self):
+    """Test that provider is successfully parsed when spec is valid."""
+    spec = {
+        'type': 'mavenJar',
+        'config': {
+            'group_id': 'org.apache.beam',
+            'artifact_id': 'test-library',
+            'version': '1.0.0'
+        },
+        'transforms': self._urns
+    }
+    provider = ExternalProvider.provider_from_spec(spec)
+    self.java_provider_from_spec_asserts(provider)
+
+
+class ExternalPythonProviderTest(unittest.TestCase):
+  """Test class for ExternalPythonProvider tests."""
+  def __init__(self, method_name="runExternalPythonProviderTest"):
+    super().__init__(method_name)
+
+    self._type = 'pythonPackage'
```

Comment:
You could have _type as a class-level attribute and not bother overriding __init__.